### ISSUE TITLE
Problems with mTLS certs

### ISSUE THREAD

**Issue by dexion** [hasura/ndc-http#41]
I can't add certificates.

I get this error.  I tried 0.5.0 and 0.6.0 versions
```
goroutine 1 [running]:
github.com/hasura/ndc-http/connector/internal/security.loadTLSConfig(0xc000220280)
        /app/connector/internal/security/tls.go:89 +0x3dc
github.com/hasura/ndc-http/connector/internal/security.NewHTTPClientTLS(0x1c6d820, 0xc000346300?)
        /app/connector/internal/security/tls.go:25 +0x65
github.com/hasura/ndc-http/connector/internal.(*UpstreamManager).Register(0xc00022d938, {0x1333ca0, 0xc000346300}, {0xc00025e060, 0x52}, 0xc0002181e0)
        /app/connector/internal/upstream.go:61 +0x8d
github.com/hasura/ndc-http/connector.(*HTTPConnector).ApplyNDCHttpSchemas(0xc0000e3860, {0x1333ca0, 0xc000346300}, 0xc000218060, {0xc000276000?, 0x7f335393d108?, 0xc0001bbb50?}, 0xc0003254a0)
        /app/connector/schema.go:32 +0x18e
github.com/hasura/ndc-http/connector.(*HTTPConnector).ParseConfiguration(0xc0000e3860, {0x1333ca0, 0xc000346300}, {0xc000348810, 0xe})
        /app/connector/connector.go:83 +0x3c5
github.com/hasura/ndc-sdk-go/connector.NewServer[...]({0x133d860?, 0xc0000e3860}, 0xc000360580, {0xc00035f100, 0x4, 0xc0001bbea0})
        /go/pkg/mod/github.com/hasura/ndc-sdk-go@v1.6.3-0.20241127025002-02d7a257e75f/connector/server.go:80 +0x3a3
github.com/hasura/ndc-sdk-go/connector.StartCustom[...]({0x132c948, 0xc0000f1080}, {0x133d860, 0xc0000e3860}, {0xc0001bbf30, 0x2, 0x0})
        /go/pkg/mod/github.com/hasura/ndc-sdk-go@v1.6.3-0.20241127025002-02d7a257e75f/connector/cli.go:69 +0x4b1
github.com/hasura/ndc-sdk-go/connector.Start[...](...)
        /go/pkg/mod/github.com/hasura/ndc-sdk-go@v1.6.3-0.20241127025002-02d7a257e75f/connector/cli.go:51
main.main()
```

I tiried both: files and base64 options in my config.

config.yaml:
```
files:
  - file: swagger.json
    spec: oas2
    timeout:
      value: 30
    retry:
      times:
        value: 0
      delay:
        value: 1000
      httpStatus: [429, 500, 502, 503]
    patchAfter:
      - path: patch-config-after.yaml
        strategy: merge
```

patch-config-after.yaml:
```
settings:
  servers:
      - url:
          env: MTLS_API_URL
  securitySchemes:
    mtls:
      type: mutualTLS
  tls:
    # Path to the TLS cert to use for TLS required connections.
    # certFile:
    #   env: MTLS_API_CERT_FILE
    
    # Alternative to cert_file. Provide the certificate contents as a base64-encoded string instead of a filepath.
    certPem:
      env: MTLS_API_CERT_PEM
    
    # Path to the TLS key to use for TLS required connections.
    # keyFile:
    #   env: MTLS_API_KEY_FILE
    
    # Alternative to key_file. Provide the key contents as a base64-encoded string instead of a filepath.
    keyPem:
      env: MTLS_API_KEY_PEM
    
    # Path to the CA cert.
    # caFile:
    #   env: MTLS_API_CA_FILE
    
    # Alternative to ca_file. Provide the CA cert contents as a base64-encoded string instead of a filepath.
    caPem:
      env: MTLS_API_CA_PEM
    
    # Additionally you can configure TLS to be enabled but skip verifying the server's certificate chain (optional).
    insecureSkipVerify:
      env: MTLS_API_INSECURE_SKIP_VERIFY
      value: true
    
    # Whether to load the system certificate authorities pool alongside the certificate authority (optional).
    # includeSystemCACertsPool:
    #   env: MTLS_API_INCLUDE_SYSTEM_CA_CERT_POOL
    #   value: false
    
    # ServerName requested by client for virtual hosting (optional).
    # serverName:
    #   env: MTLS_API_SERVER_NAME
    
    # Minimum acceptable TLS version (optional).
    # minVersion: "1.0"
    
    # Maximum acceptable TLS version (optional).
    # maxVersion: "1.3"
    
    # Explicit cipher suites can be set. If left blank, a safe default list is used (optional).
    # cipherSuites:
    #   - TLS_AES_128_GCM_SHA256
```

I used `patch-config-after.yaml` because I couldn't insert mTLS config inside basic config.yaml

**Comment by hgiasac** [comment]
Hi @dexion 

Thanks for reporting the bug. Can you change the default value of `insecureSkipVerify` to false or remove the env?

```yaml
insecureSkipVerify:
  env: MTLS_API_INSECURE_SKIP_VERIFY
  value: false
```

However, I doubt that the cert keys weren't loaded correctly. You should use inline `*_PEM` base64 if you deploy the connector to the DDN cloud.  File path env may not work in DDN cloud deployment. It would be helpful if you could share the `connector/compose.yaml` file and `.env` file with sensitive secrets masked.

**Comment by dexion** [comment]
I try to deploy all of it localy. 

Changing the default value of `insecureSkipVerify` to `false` resolves the main issue, but I can't connect to remote http-server. I got this error from connector:

```
app_mtls_api-1          | {"time":"2024-12-11T17:09:44.716642711Z","level":"ERROR","msg":"Internal Server Error","request_id":"8ae581f0-7775-4370-ba2a-56246fc525e1","latency":37118184,"request":{"method":"POST","remote_address":"192.168.65.1:25170","url":"/query"},"response":{"body":"{\"message\":\"Get \\\"https://192.168.0.100/api/users\\\": tls: failed to verify certificate: x509: certificate signed by unknown authority\",\"details\":{}}","headers":{"Content-Type":["application/json"]},"status":500}}
```

I need `MTLS_API_INSECURE_SKIP_VERIFY` because my CA is untrusted. For example, with `curl` I have to add `-k` key to skip the server's cert verification.

`connector/compose.yaml`:
```
services:
  app_mtls_api:
    build:
      context: .
      dockerfile: .hasura-connector/Dockerfile.mtls_api
    environment:
      HASURA_SERVICE_TOKEN_SECRET: $APP_MTLS_API_HASURA_SERVICE_TOKEN_SECRET
      OTEL_EXPORTER_OTLP_ENDPOINT: $APP_MTLS_API_OTEL_EXPORTER_OTLP_ENDPOINT
      OTEL_SERVICE_NAME: $APP_MTLS_API_OTEL_SERVICE_NAME
    extra_hosts:
      - local.hasura.dev:host-gateway
    ports:
      - 7324:8080
```


`.env`:
```
APP_MTLS_API_AUTHORIZATION_HEADER="Bearer v..."
APP_MTLS_API_HASURA_SERVICE_TOKEN_SECRET="v..."
APP_MTLS_API_OTEL_EXPORTER_OTLP_ENDPOINT="http://local.hasura.dev:4317"
APP_MTLS_API_OTEL_SERVICE_NAME="app_mtls_api"
APP_MTLS_API_READ_URL="http://local.hasura.dev:7324"
APP_MTLS_API_WRITE_URL="http://local.hasura.dev:7324"

MTLS_API_CA_FILE="/full_path_to/app/connector/mtls_api/ca.crt"
MTLS_API_CA_PEM="L...\n"
MTLS_API_CERT_FILE="/full_path_to/app/connector/mtls_api/client.crt"
MTLS_API_CERT_PEM="Q...\n"
MTLS_API_INSECURE_SKIP_VERIFY="false"
MTLS_API_KEY_FILE="/full_path_to/app/connector/mtls_api/client.key"
MTLS_API_KEY_PEM="L...\n"
MTLS_API_URL="https://192.168.0.100/api"
```

**Comment by hgiasac** [comment]
I see. You need to add env mapping to the `compose.yaml` file. Those mTLS env variables are customized after introspection so you need to add them manually. 

```yaml
services:
  app_mtls_api:
    build:
      context: .
      dockerfile: .hasura-connector/Dockerfile.mtls_api
    environment:
      HASURA_SERVICE_TOKEN_SECRET: $APP_MTLS_API_HASURA_SERVICE_TOKEN_SECRET
      OTEL_EXPORTER_OTLP_ENDPOINT: $APP_MTLS_API_OTEL_EXPORTER_OTLP_ENDPOINT
      OTEL_SERVICE_NAME: $APP_MTLS_API_OTEL_SERVICE_NAME
      MTLS_API_CA_PEM: $MTLS_API_CA_PEM
      MTLS_API_CERT_PEM: $MTLS_API_CERT_PEM
      MTLS_API_INSECURE_SKIP_VERIFY: $MTLS_API_INSECURE_SKIP_VERIFY
      MTLS_API_KEY_PEM: $MTLS_API_KEY_PEM
      MTLS_API_URL: $MTLS_API_URL
    extra_hosts:
      - local.hasura.dev:host-gateway
    ports:
      - 7324:8080
```

And add `envMapping` to  `app/subgraph.yaml` file as well

```yaml
envMapping:
  # ...
  MTLS_API_CA_PEM:
    fromEnv: MTLS_API_CA_PEM
  MTLS_API_CERT_PEM:
    fromEnv: MTLS_API_CERT_PEM
  MTLS_API_INSECURE_SKIP_VERIFY:
    fromEnv: MTLS_API_INSECURE_SKIP_VERIFY
  MTLS_API_KEY_PEM:
    fromEnv: MTLS_API_KEY_PEM
  MTLS_API_URL:
    fromEnv: MTLS_API_URL
```

**Comment by dexion** [comment]
It works with your additional configurations. Thank you!

**Comment by hgiasac** [comment]
@dexion the panic bug was fixed in [v0.6.1](https://github.com/hasura/ndc-http/releases/tag/v0.6.1). You can change the version in `Dockerfile.mtls_api`, `.hasura-connector/connector-metadata.yaml`, and `connector.yaml` to upgrade.
